src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -97,56 +118,130 @@ internal class PayloadSenderV2 : IPayloadSender, IDisposable
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
+			var enqueuedSuccessfully = _eventQueue.Post(eventObj);
+			if (!enqueuedSuccessfully)
+			{
+				_logger.Debug()
+					?.Log("Failed to enqueue " + dbgEventKind + "."
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
 			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Error to the queue, {Error}"
-					: "Error added to the queue, {Error}", error);
+				?.Log("Enqueued " + dbgEventKind + "."
+					+ " newEventQueueCount: {EventQueueCount}."
+					+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+					, newEventQueueCount, eventObj);
+
+			if (_flushInterval == TimeSpan.Zero) _eventQueue.TriggerBatch();
+
+			return true;
 		}
 
 		private async Task DoWork()
 		{
+			var minTimeBetweenWaitingOnTimerLogs = TimeSpan.FromSeconds(10);
+			var stopwatch = Stopwatch.StartNew();
+			TimeSpan? elapsedOnLastWaitingOnTimerLog = null;
+
 			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
+			Task<object[]> receiveAsyncTask = null;
 			while (true)
 			{
-				var queueItems = await _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
-				await ProcessQueueItems(queueItems);
+				// ReSharper disable once InconsistentlySynchronizedField
+				if (receiveAsyncTask == null) receiveAsyncTask = _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
+
+				object[] eventBatchToSend = null;
+				if (_flushInterval == TimeSpan.Zero)
+				{
+					_logger.Trace()?.Log("Waiting for data to send... (not using FlushInterval timer because FlushInterval is 0)");
+					eventBatchToSend = await receiveAsyncTask;
+					receiveAsyncTask = null;
+				}
+				else
+				{
+					var logTimerRelatedInfo = false;
+					var elapsedTime = stopwatch.Elapsed;
+					if (!elapsedOnLastWaitingOnTimerLog.HasValue ||
+						elapsedOnLastWaitingOnTimerLog.Value + minTimeBetweenWaitingOnTimerLogs <= elapsedTime)
+					{
+						logTimerRelatedInfo = true;
+						_logger.Trace()
+							?.Log("Waiting for data to send or FlushInterval timer to be triggered (whichever is earlier)..."
+								+ " _flushInterval: {FlushInterval}", _flushInterval);
+						elapsedOnLastWaitingOnTimerLog = elapsedTime;
+					}
+
+					var flushIntervalDelayTask = Task.Delay((int)_flushInterval.TotalMilliseconds, _batchBlockReceiveAsyncCts.Token);
+					var completedTask = await Task.WhenAny(receiveAsyncTask, flushIntervalDelayTask);
+					if (completedTask == receiveAsyncTask)
+					{
+						eventBatchToSend = await receiveAsyncTask;
+						receiveAsyncTask = null;
+					}
+					else
+					{
+						Assertion.IfEnabled?.That(completedTask == flushIntervalDelayTask,
```

Comment:
I've rewritten this piece of code (mainly because it was leaking `Task.Delay` timers (I've added a detailed comment - https://github.com/elastic/apm-agent-dotnet/pull/487/files#diff-886de9ff879075787e23345216866be9R231) so there's no assert there anymore.